qutebrowser/browser/webengine/notification.py

```diff
@@ -0,0 +1,134 @@
+
+    def __init__(self):
+        # Dict mapping notification IDs to the corresponding QWebEngine
+        # notification objects.
+        self.active_notifications = {}
+        bus = QDBusConnection.sessionBus()
+        if not bus.isConnected():
+            raise DBusException("Failed to connect to DBus session bus")
+
+        self.interface = QDBusInterface(
+            "org.freedesktop.Notifications",  # service
+            "/org/freedesktop/Notifications",  # path
+            "org.freedesktop.Notifications",  # interface
+            bus,
+        )
+
+        if not self.interface:
+            raise DBusException("Could not construct a DBus interface")
+
+    def set_as_presenter_for(self, profile: QWebEngineProfile) -> None:
+        """Sets the profile to use the manager as the presenter."""
+        # PyQtWebEngine unrefs the callback after it's called, for some reason.
+        # So we call setNotificationPresenter again to *increase* its refcount
+        # to prevent it from getting GC'd. Otherwise, random methods start
+        # getting called with the notification as `self`, or segfaults happen,
+        # or other badness.
+        def _present_and_reset(qt_notification: QWebEngineNotification) -> None:
+            profile.setNotificationPresenter(_present_and_reset)
+            self._present(qt_notification)
+
+        profile.setNotificationPresenter(_present_and_reset)
+
+    def _present(self, qt_notification: QWebEngineNotification) -> None:
+        """Shows a notification over DBus.
+
+        This should *not* be directly passed to setNotificationPresenter
+        because of a bug in the PyQtWebEngine bindings.
+        """
+        # notification id 0 means 'assign us the ID'. We can't just pass 0
+        # because it won't get sent as the right type.
+        notification_id = QVariant(0)
+        notification_id.convert(QVariant.UInt)
+
+        actions_list = QDBusArgument([], QMetaType.QStringList)
+
+        qt_notification.show()
+        hints: Dict[str, Any] = {
+            # Include the origin in case the user wants to do different things
+            # with different origin's notifications.
+            "x-qutebrowser-origin": qt_notification.origin().toDisplayString()
+        }
+        if not qt_notification.icon().isNull():
+            hints["image-data"] = self._convert_image(qt_notification.icon())
+
+        reply = self.interface.call(
+            QDBus.BlockWithGui,
+            "Notify",
+            "qutebrowser",  # application name
+            notification_id,
+            "qutebrowser",  # icon
+            qt_notification.title(),
+            qt_notification.message(),
+            actions_list,
+            hints,
+            -1,  # timeout; -1 means 'use default'
+        )
+
+        if not (len(reply.arguments()) == 1):
+            raise DBusException(
+                "Got an unexpected number of reply arguments {}".format(
```

Comment:
```suggestion
                "Got an unexpected number of reply arguments: {}".format(
```

(though maybe it should print the actual arguments instead, to make it easier to debug this?

Reply:
Done. We also now just check the signature insted (since that checks both length and type).

There's actually a way to autogenerate service-specific bindings, but that generates C++ code which would then need to be wrapped.